File: balancedBinaryTree.py
# Given a binary tree, determine if it is height-balanced
# A height-balanced binary tree is a binary tree in which the depth of the two subtrees of every node never differs by more than one.

class Solution(object):
    def getDepth(self, root):
        if not root:
            return 0
        
        leftDepth = self.getDepth(root.left)
        rightDepth = self.getDepth(root.right)
        
        return max(leftDepth, rightDepth) + 1

    def isBalanced(self, root):
        if not root:
            return True
        
        leftDepth = self.getDepth(root.left)
        rightDepth = self.getDepth(root.right)

        if abs(leftDepth - rightDepth) > 1:
            return False
        else:
            return self.isBalanced(root.left) and self.isBalanced(root.right)

# Explanation
# getDepth function from previous code
# Base case to check if root is empty, return TRUE since an empty tree is balanced
# Recursively calculated the depth of the left and right subtree
# If the absolute value of leftDepth - rightDepth is greater than 1, return false as the tree is not balanced
# Otherwise recursively check if both the left and right subtrees are balanced, eventually it will evaluate to false if it is not balanced
# If both subtrees evaluate to true, the tree is balanced

File: balancedBinaryTree.c
// Given a binary tree, determine if it is height-balanced
// A height-balanced binary tree is a binary tree in which the depth of the two subtrees of every node never differs by more than one.

int maxDepth(struct TreeNode* root){
    if (root == NULL){
        return 0;
    }

    int leftDepth = maxDepth(root->left);
    int rightDepth = maxDepth(root->right);

    if (leftDepth > rightDepth){
        return leftDepth + 1;
    }
    else{
        return rightDepth + 1;
    }
}

bool isBalanced(struct TreeNode* root){
    if (root == NULL){
        return true;
    }

    int depthLeft = maxDepth(root->left);
    int depthRight = maxDepth(root->right);

    if (abs(depthLeft - depthRight) > 1){
        return false;
    }
    else{
        return isBalanced(root->left) && isBalanced(root->right);
    }
}

// Explanation
// maxDepth function from previous code
// Base case to check if tree is empty, which is a balanced tree so we return true
// Recursively calculate the depth of the left and right subtrees
// If the absolute value of the difference between of trees is > 1, return false as the tree is not balanced
// Otherwise recursively check all nodes subtrees to be balanced, eventually it will return false if the tree is not balanced
// If both return true, the tree is balanced
